fixtures/asset-config/test-cases/encoding-test-cases.ts

```diff
@@ -0,0 +1,335 @@
+import { TestCase } from "../html-handling.test";
+
+export const encodingTestCases: {
+	html_handling:
+		| "auto-trailing-slash"
+		| "drop-trailing-slash"
+		| "force-trailing-slash"
+		| "none";
+	cases: TestCase[];
+}[] = [
+	{
+		html_handling: "auto-trailing-slash",
+		cases: [
+			{
+				title: "/[boop] -> /%5Bboop%5D 307 (with /[boop].html)",
+				files: ["/[boop].html"],
+				requestPath: "/[boop]",
+				matchedFile: "/[boop].html",
+				finalPath: "/%5Bboop%5D",
+			},
+			{
+				title: "/%5Bboop%5D -> 200 (with /[boop].html)",
+				files: ["/[boop].html"],
+				requestPath: "/%5Bboop%5D",
+				matchedFile: "/[boop].html",
+				finalPath: "/%5Bboop%5D",
+			},
+			{
+				title: "/%5Bboop%5D -> 200 (with /%5Bboop%5D.html)",
+				files: ["/%5Bboop%5D.html"],
+				requestPath: "/%5Bboop%5D",
+				matchedFile: "/%5Bboop%5D.html",
+				finalPath: "/%5Bboop%5D",
+			},
+			{
+				title: "/[boop] -> 307 (with /%5Bboop%5D.html)",
+				files: ["/%5Bboop%5D.html"],
+				requestPath: "/[boop]",
+				matchedFile: "/%5Bboop%5D.html",
+				finalPath: "/%5Bboop%5D",
+			},
+			{
+				title: "/%5Bboop%5D.html -> 307 (with /%5Bboop%5D.html)",
+				files: ["/%5Bboop%5D.html"],
+				requestPath: "/%5Bboop%5D.html",
+				matchedFile: "/%5Bboop%5D.html",
+				finalPath: "/%5Bboop%5D",
+			},
+			// mix of encoded and unencoded paths
+			{
+				title: "/beep/[b%C3%B2op] -> /beep/%5Bb%C3%B2op%5D 307",
+				files: ["/beep/[bòop].html"],
+				requestPath: "/beep/[b%C3%B2op]",
+				matchedFile: "/beep/[bòop].html",
+				finalPath: "/beep/%5Bb%C3%B2op%5D",
+			},
+			// exact matches should win
+			{
+				title: "/[boop] -> 200 (with /[boop].html)",
+				files: ["/%5Bboop%5D.html", "/[boop].html"],
```

Comment:
> If you have time, maybe cook up another version which does "/%5Bboop%5D.html at /%255Bboop%255D only" and we can evaluate if we want to make that change separately? I think it might allow you to simplify the asset serving logic a fair bit — it makes us more consistent and deterministic. 

I actually tried this out the other day, and indeed to it does make the logic a lot simpler. I've chucked that on a branch here: https://github.com/cloudflare/workers-sdk/tree/double-encode